### PR TITLE
Remove unnecessary uses of `DashMap` and `Arc`

### PR DESCRIPTION
**Pull request by ibraheemdev**
## Summary

All of the resolver code is run on the main thread, so a lot of the `Send` bounds and uses of `DashMap` and `Arc` are unnecessary. We could also switch to using single-threaded versions of `Mutex` and `Notify` in some places, but there isn't really a crate that provides those I would be comfortable with using.

The `Arc` in `OnceMap` can't easily be removed because of the uv-auth code which uses the [reqwest-middleware](https://docs.rs/reqwest-middleware/latest/reqwest_middleware/trait.Middleware.html) crate, that seems to adds unnecessary `Send` bounds because of `async-trait`. We could duplicate the code and create a `OnceMapLocal` variant, but I don't feel that's worth it.

### CODE REVIEW

File: crates/uv-resolver/src/resolver/mod.rs
Comment (ibraheemdev) [inline crates/uv-resolver/src/resolver/mod.rs:142]: :relieved:

File: crates/uv-resolver/src/resolver/mod.rs
Comment (ibraheemdev) [inline crates/uv-resolver/src/resolver/mod.rs:144]: :relieved:

**Comment by BurntSushi** [comment]
> All of the resolver code is run on the main thread, so a lot of the `Send` bounds and uses of `DashMap` and `Arc` are unnecessary.

Is the resolver running on just the main thread what we want? If we wanted it to make use of multiple threads in the future (even as an experiment to try?), then adding these `Send` bounds back could be pretty annoying. (I'm not sure our uses of `DashMap` and `Arc` are costing us much, but maybe I'm wrong about that. Of course, if we don't need them, then I agree with this PR.)

File: crates/uv-resolver/src/resolver/mod.rs
Comment (charliermarsh) [reply crates/uv-resolver/src/resolver/mod.rs:144]: Love that!

**Review by charliermarsh — APPROVED** [review]
I'm supportive of the change, though good to get @BurntSushi sign-off too before merging since he chimed in with comments.

**Review by BurntSushi — APPROVED** [review]
This LGTM. I'm overall in favor of simplifying the code as it exists under current constraints, and if we find ourselves needing to add this (or something like it) back, then it doesn't seem that horrible to do.

**Pull request merged**
